src/mlpack/methods/bias_svd/bias_svd_function_impl.hpp

```diff
@@ -0,0 +1,411 @@
+double BiasSVDFunction<MatType>::Evaluate(const arma::mat& parameters,
+                                                 const size_t start,
+                                                 const size_t batchSize) const
+{
+  // It's possible this loop could be changed so that it's SIMD-vectorized.
+  double objective = 0.0;
+  for (size_t i = start; i < start + batchSize; ++i)
+  {
+    // Indices for accessing the the correct parameter columns.
+    const size_t user = data(0, i);
+    const size_t item = data(1, i) + numUsers;
+
+    // Calculate the squared error in the prediction.
+    const double rating = data(2, i);
+    const double userBias = parameters(rank, user);
+    const double itemBias = parameters(rank, item);
+    double ratingError = rating - userBias - itemBias -
+        arma::dot(parameters.col(user).subvec(0, rank - 1),
+                  parameters.col(item).subvec(0, rank - 1));
+    double ratingErrorSquared = ratingError * ratingError;
+
+    // Calculate the regularization penalty corresponding to the parameters.
+    double userVecNorm = arma::norm(parameters.col(user), 2);
+    double itemVecNorm = arma::norm(parameters.col(item), 2);
+    double regularizationError = lambda * (userVecNorm * userVecNorm +
+                                           itemVecNorm * itemVecNorm);
+
+    objective += (ratingErrorSquared + regularizationError);
+  }
+
+  return objective;
+}
+
+template <typename MatType>
+void BiasSVDFunction<MatType>::Gradient(const arma::mat& parameters,
+                                               arma::mat& gradient) const
+{
+  // For an example with rating corresponding to user 'i' and item 'j', the
+  // gradients for the parameters is as follows:
+  //           grad(u(i)) = 2 * (lambda * u(i) - error * v(j))
+  //           grad(v(j)) = 2 * (lambda * v(j) - error * u(i))
+  //           grad(p(i)) = 2 * (lambda * p(i) - error)
+  //           grad(q(j)) = 2 * (lambda * q(j) - error)
+  // 'error' is the prediction error for that example, which is:
+  //           rating(i, j) - p(i) - q(j) - u(i).t() * v(j)
+  // The full gradient is calculated by summing the contributions over all the
+  // training examples.
+
+  gradient.zeros(rank + 1, numUsers + numItems);
+
+  for (size_t i = 0; i < data.n_cols; i++)
+  {
+    // Indices for accessing the the correct parameter columns.
+    const size_t user = data(0, i);
+    const size_t item = data(1, i) + numUsers;
+
+    // Prediction error for the example.
+    const double rating = data(2, i);
+    const double userBias = parameters(rank, user);
+    const double itemBias = parameters(rank, item);
+    double ratingError = rating - userBias - itemBias -
+        arma::dot(parameters.col(user).subvec(0, rank - 1),
+                  parameters.col(item).subvec(0, rank - 1));
+
+    // Gradient is non-zero only for the parameter columns corresponding to the
+    // example.
+    gradient.col(user).subvec(0, rank - 1) +=
```

Comment:
Looks like the only reason of the function is vectorized sums. Do we really need this? I suggest to replace the body of the function by
```
{
  Gradient(parameters, 0, gradient, data.n_cols);
}
```

Reply:
I am not totally sure but I think vectorized sum can be faster on certain hardwares that supports SIMD.